### DISCUSSION TITLE
Force longhorn on RKE2 servers only, not on RKE2 agents

### DISCUSSION THREAD

**Discussion by dchiaramello** [longhorn/longhorn#9137] [Q&A] [answered]
Hello.

RKE2 version: "**v1.28.11+rke2r1**"
longhorn version: "**v1.6.2**"

We plan to have an RKE2 cluster with 3 `server` nodes and some "specialized" `agent` ones, in an Airgap environment.
We would like to use `docker.io/registry` configured with `longhorn` for its storage on the `server` nodes (the only ones large enough for all our Docker images).

But I failed to configure `longhorn` so that its is deployed only on the `server` nodes...

Note that I'm not an expert in Kubernetes, that project is a kind of initiation for me: I learned many things recently, but have much more to learn... So it's possible that the way I'm trying to do that is bad!

I have configured the default `longhorn` Manifest (before it is deployed, after a "from scratch" RKE2 cluster deployment) with `create-default-disk-labeled-nodes: true` declared in its `default-setting.yaml` ConfigMap, and labeled all my `server` nodes with a `node.longhorn.io/create-default-disk=true` `node-label`, hoping that it would do the trick - but I notice that many PODs are attempted to be created on `agent` nodes, and are failing as expected:
- 2 instances of `longhorn-ui` are in `Error` status  because of "_The node was low on resource: ephemeral-storage_"
- same for one instance of `longhorn-driver-deployer` in `ContainerStatusUnknown` status
- some instances of `longhorn-csi-plugin` and `engine-image-ei` are in `Evicted` status because of `DiskPressure` condition

Did I miss something? Are there additional things to configure?

Note that initially, my 'server' nodes were also using a `CriticalAddonsOnly=true:NoExecute` taint, but the `longhorn` PODs were failing to deploy on these nodes because of `untolerated taint {CriticalAddonsOnly: true}`, despite the fact that I added `taint-toleration: "CriticalAddonsOnly=true:NoSchedule"` in its `default-setting.yaml` ConfigMap - but that's another story...

**Comment by ChanYiLin** [comment, marked as answer]
Hi, if you want to deploy Longhorn only on part of the nodes in the cluster, you can set the node selector value in the helm value when deploying Longhorn.
Here is the document for the setting: https://longhorn.io/docs/1.6.2/advanced-resources/deploy/node-selector/

**Reply by dchiaramello** [reply]
Hello.

I noticed the create-default-disk-labeled-nodes attribute, but I missed the system-managed-components-node-selector one...
I will test that, thanks a lot!

**Reply by dchiaramello** [reply]
Hello again.

Before applying the Manifest with `kubectl`, I'm configuring it as follow:
```
# Source: longhorn/templates/default-setting.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: longhorn-default-setting
  namespace: longhorn-system
  labels:
    app.kubernetes.io/name: longhorn
    app.kubernetes.io/instance: longhorn
    app.kubernetes.io/version: v1.6.2
data:
  default-setting.yaml: |-
    priority-class: longhorn-critical
    create-default-disk-labeled-nodes: true
    system-managed-components-node-selector: "application.layer:master"
```
But after I have applied the Manifest, while the `create-default-disk-labeled-nodes` setting has effectively been set to `true`, the `system-managed-components-node-selector` one is still empty:
> `kubectl get settings system-managed-components-node-selector -n longhorn-system -o json`
```
{
    "apiVersion": "longhorn.io/v1beta2",
    "kind": "Setting",
    "metadata": {
        "annotations": {
            "longhorn.io/configmap-resource-version": "2146"
        },
        "creationTimestamp": "2024-08-02T09:07:39Z",
        "generation": 1,
        "name": "system-managed-components-node-selector",
        "namespace": "longhorn-system",
        "resourceVersion": "2273",
        "uid": "bc7041b0-1307-4970-9140-2121f257cdbe"
    },
    "value": ""
}
```
I continue to investigate...

**Reply by dchiaramello** [reply]
Please forget my previous message, about `system-managed-components-node-selector` being empty...
Initially, I was replacing the ConfigMap in a second step of my longhorn configuration with a personalized one, before I discoreved that this new ConfigMap was applied only on new nodes... But I forgot to remove that step in my build, so the "initial" ConfigMap that I updated following your advices was overwritten immediately after with my "personalized" ConfigMap, which was not based on your advices!!!

So, all is well, it's now behaving as expected, thanks a lot.
